crac/src/main/java/io/micronaut/crac/CracEventPublisher.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2017-2022 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.crac;
+
+import io.micronaut.context.event.ApplicationEventPublisher;
+import io.micronaut.crac.events.AfterRestoreEvent;
+import io.micronaut.crac.events.BeforeCheckpointEvent;
+import io.micronaut.runtime.context.scope.refresh.RefreshEvent;
+import jakarta.inject.Singleton;
+
+import java.util.Optional;
+import java.util.function.LongSupplier;
+
+/**
+ * A helper bean to assist in publishing the correct CRaC events when a checkpoint is taken or restored.
+ *
+ * @author Tim Yates
+ * @since 1.0.0
+ */
+@Singleton
+public class CracEventPublisher {
+
+    private final ApplicationEventPublisher<BeforeCheckpointEvent> beforeCheckpointEventPublisher;
+    private final ApplicationEventPublisher<AfterRestoreEvent> afterRestoreEventPublisher;
+    private final Optional<ApplicationEventPublisher<RefreshEvent>> optionalRefreshEventPublisher;
```

Comment:
Typically, we never use `Optional` for private instance variables. 

Please replace it with: 

```
@Nullable
 private final ApplicationEventPublisher<RefreshEvent> optionalRefreshEventPublisher;
```